Tutorial on how to create a 3Box chatbox plugin (#119)
kauri-io/Content: A tutorial on how to create a 3Box chatbox plugin.md

 5.How to build with 3box chatbox plugins
 6.Reasons for using 3box chatbox plugins

A TUTORIAL ON HOW TO CREATE A 3BOX CHATBOX PLUGIN

> # INTRODUCTION TO 3BOX CHATBOX PLUGIN

Reviewer: ```suggestion
# Introduction to 3box Chatbox Plugin
```

## What is Chatbox

> Chatbox is a 3Box plugin which enables dapp developers to add chatrooms to their react Ethereum application with a few lines of code to enable rich social discourse. The 3Box Chatbox plugin is built using 3Box infrastructure, 3Box Ghost Threads and handles all 3Box and web3 logic for creating a chat box. 3Box is creating a rich plugin ecosystem that simplifies the web3 development experience.

Reviewer: ```suggestion
Chatbox is a [3Box plugin](https://docs.3box.io/build/plugins) which enables dapp developers to add chatrooms to their react Ethereum application with a few lines of code to enable rich social discourse. The 3Box Chatbox plugin is built using 3Box infrastructure, [3Box Ghost Threads](https://docs.3box.io/build/web-apps/messaging/ghost-threads) and handles all 3Box and web3 logic for creating a chat box.
```

> ## Ghost threads

Reviewer: ```suggestion
### Ghost threads
```

> Ghost threads are a different type of messaging thread, that do not persist messages to OrbitDB database storage. Instead, ghost messages are sent from one peer to other peers on the network using IPFS pubsub, and are kept in-memory upon receipt by online peers. New users coming online or joining the chatroom can request a message backlog from other users in the chat. The Ghost Threads API allows Ethereum app developers to create new chat features that demand scalable performance to support many users sending many messages such as site-wide chat rooms, lightning fast load speeds, or novel properties such as disappearing messages.

Reviewer: ```suggestion
Ghost threads are a different type of messaging thread that do not persist messages to OrbitDB database storage. Instead, ghost messages are sent from one peer to other peers on the network using IPFS pubsub, and are kept in-memory upon receipt by online peers. New users coming online or joining the chatroom can request a message backlog from other users in the chat. The Ghost Threads API allows Ethereum app developers to create new chat features that demand scalable performance to support many users sending many messages such as site-wide chat rooms, lightning fast load speeds, or novel properties such as disappearing messages.
```

> Ghost threads are a different type of messaging thread, that do not persist messages to OrbitDB database storage. Instead, ghost messages are sent from one peer to other peers on the network using IPFS pubsub, and are kept in-memory upon receipt by online peers. New users coming online or joining the chatroom can request a message backlog from other users in the chat. The Ghost Threads API allows Ethereum app developers to create new chat features that demand scalable performance to support many users sending many messages such as site-wide chat rooms, lightning fast load speeds, or novel properties such as disappearing messages.

Reviewer: @writeprovidence This whole section is mostly just a copy and paste from the 3box docs, so probably best just to link to them instead and remove this, as I suggest in the suggestion above.

> # USE CASES OF 3BOX CHATBOX PLUGIN

Reviewer: ```suggestion
## Use Cases of 3box Chatbox Plugin
```

> Asset-driven use cases are a great start, and native to blockchains, most activity on mainstream applications today is more social and less financial in nature. Chatbox provide simple social functionality that drives everyday engagement on dapps Below are some available use cases.

Reviewer: ```suggestion
Asset-driven use cases are a great start, and native to blockchains, but most activity on mainstream applications today is more social and less financial in nature. Chatbox provide simple social functionality that drives everyday engagement on dapps Below are some available use cases.
```

> Asset-driven use cases are a great start, and native to blockchains, most activity on mainstream applications today is more social and less financial in nature. Chatbox provide simple social functionality that drives everyday engagement on dapps Below are some available use cases.

Reviewer: @writeprovidence I don't see use cases below, but implementation examples

> ## Decide on a pop-up or in-page component.

Reviewer: ```suggestion
### Decide on a pop-up or in-page component.
```

Chatbox can configured to be displayed as a pop-up modal or as an embedded in-page element depending on the user experience you wish to provide.

> ## Swap in your app’s name, image/logo, and color.

Reviewer: ```suggestion
### Swap in your app’s name, image/logo, and color.
```

> Chatbox, with its simple drop-in functionality, makes it easier than ever to experiment with social web3 development. it’s fully-featured out-of-the-box and can scale with your application into production.

Reviewer: ```suggestion
Chatbox, with its simple drop-in functionality, makes it easier to experiment with social web3 development. it’s fully-featured out-of-the-box and can scale with your application into production.
```

> # HOW 3BOX CHATBOT WORKS

Reviewer: ```suggestion
## How 3box Chatbox Works
```

> ## Architecture

Reviewer: ```suggestion
### Architecture
```

> ## Architecture

Reviewer: @writeprovidence This is a just a copy and paste from the 3box docs, we don't need to duplicate, either add your own take, or just link to their docs instead.

The Chatbox plugin is built using a standard implementation of 3Box Ghost Threads which are defined in the 3Box Threads API and made available via the 3Box.js SDK. As with all Ghost Threads, Chatbox messages are sent from one peer to other peers presently connected to the network using IPFS/libp2p pubsub and are then stored in-memory by online peers. The message backlog is persisted as long as there is at least one user in the chatbox, however if all users go offline the history will disappear. 
The Chatbox plugin includes UI for an embedded in-window or pop-up chat room along with all relevant logic. The component is configurable to various authentication patterns, and can handle both Web3/3Box logged-in and logged-out states.

> ## Authentication

Reviewer: ```suggestion
### Authentication
```

> ## Authentication

Reviewer: @writeprovidence Same as above, this is mostly just a copy and paste, either add your own thoughts, or just link to this section

Chatbox messages cannot be read until a user has authenticated their 3Box, authenticated the app's Space, and joined the Chatbox's Ghost Thread. After authenticating and joining, a user can post and receive messages from other users in real time.

> # EXAMPLE OF 3BOX CHATBOX PLUGIN

Reviewer: ```suggestion
## Example of 3box Chatbox Plugin
```

> # EXAMPLE OF 3BOX CHATBOX PLUGIN

Reviewer: @writeprovidence If this is the example use case, then move to above where you mention that you will list use cases

To make web3 a more social and interactive place, the ability for users to leave comments on 3Box profiles. This feature is known as 3Box Wall, and it opens up a whole set of new possibilities for social interaction between web3 accounts. And best of all, comments are completely decentralized since they’re built using the 3Box Threads API provided by the 3Box.js SDK.

> # HOW TO BUILD WITH 3BOX CHATBOX PLUGINS

Reviewer: ```suggestion
## How to Build with 3box Chatbox Plugins
```

> Install the Chatbox component

Reviewer: ```suggestion
```

> Choose your authentication pattern

Reviewer: ```suggestion
```

> Configure applications settings

Reviewer: ```suggestion
```

> Usage

Reviewer: ```suggestion
```

> ## Install the Chatbox component

Reviewer: ```suggestion
### Install the Chatbox component
```

> npm i -S 3box-chatbox-react

Reviewer: ```suggestion
```shell
mkdir 3box-chatbox-demo
cd 3box-chatbox-demo
npm i -S 3box-chatbox-react
```
```

> ## Choose your authentication pattern

Reviewer: ```suggestion
### Choose your authentication pattern
```
